### PR TITLE
homework 6

### PR DESCRIPTION
**Pull request by FuturrCoder**
Reorganize files, add a custom hook, and create generic components

### CODE REVIEW

**Comment by FuturrCoder** [comment]
I couldn't think of a way to use `useRef`, `useMemo`, `useCallback`, or `React.memo` to improve performance (but I do understand how they work)

**Comment by FuturrCoder** [comment]
The first three commits are from homework 5.

File: src/components/H2.js
Comment (amyzliao) [inline src/components/H2.js:1]: i would recommend combining H1 and H2 into the same file. call it `StyledTags.js` or something.
You can export multiple things from the same file by using the `export` keyword when defining a thing:
```js
// from within StyledTags.js
export function H1() ...
export function H2() ...
```
And then when you import it you can destructure like this
```js
// from within Classes.js or other files that use H1 or h2
import { H1, H2 } from "src/components/StyledTags.js"
```

Comment (amyzliao) [reply src/components/H2.js:1]: and low key you can make the component names more descriptive, like instead of calling it "H1" you can call it "Title" and instead of "H2" use "Subtitle" or something like that

File: src/App.js [hunk 1]
```diff
@@ -1,79 +1,19 @@
...
-            classes.map((c) => (
-              <ClassCard {...c} handleRemove={() => handleRemove(c.id)} key={c.id}/>
-            ))}
-        </div>
-      </div>
-      <Footer/>
-    </>
+    <BrowserRouter>
+      <Routes>
+        <Route element={<PageLayout/>}>
```
Comment (amyzliao) [inline src/App.js:12]: i LOVE your use of layout components here!!!!

File: src/App.js [hunk 2]
```diff
@@ -1,79 +1,19 @@
...
-    title: "CONDUCT 393-0",
-    description: "Orchestral Organizations",
-    members: 77,
-    url: "/classes/conduct-393-0"
-  }
-];
+import {BrowserRouter, Route, Routes} from "react-router-dom";
+import PageLayout from "./navigation/PageLayout";
+import Classes from "./classes/Classes";
+import Class from "./classes/class/Class";
```
Comment (amyzliao) [inline src/App.js]: since `Class` is an entirely different page from `Classes`, I would recommend putting them on the same directory level instead of having `Class` be a subdirectory of `Classes`.
my recommendation is to have a src/pages/ directory, with src/pages/classes and src/pages/class subdirectories.
src/pages/classes would contain `Classes.js` and other related files
src/pages/class would contain `Class.js` and other related files

File: src/hooks/useUsers.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import {useEffect, useState} from "react";
+
```
Comment (amyzliao) [inline src/hooks/useUsers.js:2]: a fantastic use of custom hooks!!! love it!!!!

File: src/classes/class/Class.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+      <p className="mt-4 text-lg text-neutral-gray">{classData.members} members</p>
+      <div className="mt-2 flex flex-nowrap gap-3 overflow-x-auto">
+        {users.length === 0 ?
+          <p className="mt-4 test-md text-neutral-gray">{message}</p> :
+          users.map((user) => (
+            <PersonCard {...user} key={user.id}/>
+          ))
+        }
+      </div>
+    </div>
```
Comment (amyzliao) [inline src/classes/class/Class.js]: I recommend making more child components to make your code more readable and clean up this component a bit more.
For high-level components with lots of children, especially if the children have different purposes/focuses, hiding away implementation details in other components is really important and useful.
For example, your `Class` component can be summarized by "a title, a description, a member count, and list of users".
your return statement would thus look like:
```js
return (
  <div>
    <Title/>
    <Description />
    <MemberCount />
    <UsersList />
  </div>
)
```
This way, it's a lot cleaner and easier for someone to get a rough idea of what's inside this component. If they then wanted more details about the `UsersList`, they could find the `UsersList` component defined in `UsersList.js` and get all the details they need.
In this way, your `Class` component acts more like a table of contents than the contents themselves.

File: src/hooks/useUsers.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+const useUsers = () => {
+  const [users, setUsers] = useState([]);
+  const [message, setMessage] = useState("");
```
Comment (amyzliao) [inline src/hooks/useUsers.js:5]: i like the fact that you are doing error handling here! its always important to think about "what should i display if things go wrong?" because things will go wrong all the time haha

File: src/classes/class/Class.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+import useUsers from "../../hooks/useUsers";
+import H1 from "../../components/H1";
+
+function Class() {
+  const {classURL} = useParams();
+
+  // if CLASSES was large, then this would be expensive so useMemo would be useful
+  const classData = useMemo(() => {
+    return CLASSES.find(c => c.url === classURL)
+  }, [classURL]);
```
Comment (amyzliao) [inline src/classes/class/Class.js]: from what i can tell, the only thing that could trigger a rerender of `Class` is if `classURL` changes. (or if `users` or `message` change, but that will only happen once during the initial render). 
this means that this `useMemo` doesn't really do anything because every time `Class` rerenders, it will be because `classURL` changes, and then we would run the `useMemo` computation here. it's not saving us any work.
if, however, `Class` was dealing with other states unrelated to the classURL, and those states might be frequently changing, then `useMemo` would make sense here.

basically, the point of `useMemo` is what is *excluded* from the dependency array -- we are trying to avoid rerendering when those excluced items change -- which means there have to be reactive variables that change a lot that are *excluded* from the dependency array for `useMemo` to have any effect.

File: src/classes/class/Class.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+  return (
+    <div className="px-8 py-10">
+      <H1>{classData.title}</H1>
+      <p className="mt-3 text-md text-neutral-gray">{classData.description}</p>
+      <p className="mt-4 text-lg text-neutral-gray">{classData.members} members</p>
+      <div className="mt-2 flex flex-nowrap gap-3 overflow-x-auto">
+        {users.length === 0 ?
+          <p className="mt-4 test-md text-neutral-gray">{message}</p> :
+          users.map((user) => (
+            <PersonCard {...user} key={user.id}/>
```
Comment (amyzliao) [inline src/classes/class/Class.js]: excellent use of map with keys!!!

File: src/classes/class/PersonCard.js [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+  return (
+    <div className="px-4 py-8 w-36 flex-none border border-gray-200 rounded-2xl flex flex-col gap-3 items-center">
+      <img src={props.profilePicture} alt={`${props.firstName}'s avatar`}
+           className={`w-14 h-14 rounded-full ${loaded ? "" : "hidden"}`}
+           onError={onImageError} onLoad={onImageLoad}/>
+      <FontAwesomeIcon icon={faCircleUser} clipPath="inset(0% round 999px)"
+                       className={`w-14 h-14 text-neutral-gray ${loaded ? "hidden" : ""}`}/>
+      <h3 className="text-neutral-black text-sm text-center w-full overflow-hidden overflow-ellipsis whitespace-nowrap">
+        {props.firstName} {props.lastName}
+      </h3>
```
Comment (amyzliao) [inline src/classes/class/PersonCard.js]: Similar comment as `Class.js` about splitting this up into multple components to make it more readable

File: src/config/navbar.config.js
Comment (amyzliao) [inline src/config/navbar.config.js:1]: Generally, `.config` files are used for like, environment configuration and like ugly stuff like that. 
i think you should put this in a file called `constants.js` that lives in the `navigation` folder.

File: src/classes/Classes.js [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+  return (
+    <div className="px-8 py-10">
+      <H1>Classes</H1>
+      <div className="mt-8 grid grid-cols-1 sm:grid-cols-2 lg:grid-cols-3 gap-6 items-stretch">
+        {classes.length === 0 ?
+          <p>No classes available.</p> :
+          classes.map((c) => (
+            <ClassCard {...c} handleRemove={() => handleRemove(c.id)} key={c.id}/>
+          ))}
+      </div>
```
Comment (amyzliao) [inline src/classes/Classes.js]: similar comment about splitting up into multple components like in `Class.js`

**Comment by amyzliao** [comment]
Great job! I think your component and file organization has improved dramatically ⭐
Most of my comments focus on how you can make your code even more readable and maintainable.
As far as performance goes, memoization becomes useful once you have a lot of states that change frequently in your app. Right now, you dont have too many changing states, so it wont make much of a difference, but as your app expands you'll be able to find far more use cases.

**Comment by amyzliao** [comment]
Feel free to merge this branch into main
